openfpga/src/fpga_bitstream/write_text_fabric_bitstream.cpp

```diff
@@ -245,6 +245,142 @@ static int write_memory_bank_flatten_fabric_bitstream_to_text_file(
   return status;
 }
 
+/********************************************************************
+ * Write the fabric bitstream fitting a memory bank protocol
+ * to a plain text file in efficient method
+ *
+ * Old function is write_memory_bank_flatten_fabric_bitstream_to_text_file()
+ *
+ * As compared to original function, based on 100K LE FPGA:
+ *  1. Original function used 600 seconds and need 80G Bytes of memory
+ *  2. This new function only needs 1 second and 4M Bytes
+ *
+ * Old function only print WL in decremental order. It is not by intentional
+ * It is because of the map-key ordering
+ * In QL Memory Bank with Flatten BL/WL, data is stored by WL address,
+ *   where we use WL string as map key
+ *     WL #0 --- "1000000000000 .... 0000"
+ *     WL #1 --- "0100000000000 .... 0000"
+ *     WL #n-1 --- "0000000000000 .... 0001
+ * From string comparison wise, WL #n-1 will be first, and WL #0 will be last
+ * The sequence of WL does not really matter,  but preferrable in some ordering
+ *   manner. Using map key as ordering cannot guarantee the determinstic
+ *
+ * This new way of writting fabric guarantee the WL order in 100% deterministc
+ *   way: either incremental (default) or decremental
+ *
+ * Return:
+ *  - 0 if succeed
+ *  - 1 if critical errors occured
+ *******************************************************************/
+static int fast_write_memory_bank_flatten_fabric_bitstream_to_text_file(
+  std::fstream& fp, const bool& fast_configuration,
+  const bool& bit_value_to_skip, const FabricBitstream& fabric_bitstream,
+  const bool& keep_dont_care_bits, const bool& wl_incremental_order) {
+  int status = 0;
+
+  std::string dont_care_bit = "0";
+  if (keep_dont_care_bits) {
+    dont_care_bit = "x";
+  }
+  const FabricBitstreamMemoryBank* memory_bank =
+    fabric_bitstream.memory_bank_info();
+
+  // Must call this to prepare wls_to_skip
+  (const_cast<FabricBitstreamMemoryBank*>(memory_bank))
+    ->fast_configuration(fast_configuration, bit_value_to_skip);
+
+  fabric_size_t lontest_effective_wl_addr_size =
+    memory_bank->get_lontest_effective_wl_addr_size();
+  /* Output information about how to intepret the bitstream */
+  fp << "// Bitstream length: " << lontest_effective_wl_addr_size << std::endl;
+  fp << "// Bitstream width (LSB -> MSB): ";
+  fp << "<bl_address " << memory_bank->get_total_bl_addr_size() << " bits>";
+  fp << "<wl_address " << memory_bank->get_total_wl_addr_size() << " bits>";
+  fp << std::endl;
+
+  std::vector<fabric_size_t> wl_indexes;
+  for (size_t region = 0; region < memory_bank->datas.size(); region++) {
+    if (wl_incremental_order) {
+      wl_indexes.push_back(0);
```

Comment:
Suggest to leave comments to explain how the wl_indices are organized.

Reply:
Address via https://github.com/chungshien/OpenFPGA/commit/838cf0d818166d18a51c67d3fc605b990a38e570